crowbar_framework/app/models/service_object.rb

```diff
@@ -1188,6 +1192,9 @@ def apply_role(role, inst, in_queue, bootstrap = false)
               next
             end
 
+            # skip adding nodes to the batch unless they are really in the list to be deployed
+            next unless new_elements[role_name].include?(node_name)
```

Comment:
Can you move this to either:

 - a few lines earlier, before `pre_cached_nodes[node_name] ||= Node.find_by_name(node_name)` (to avoid loading data from chef without a good reason)
 - or after changing `pending_node_actions` (since at that point, we have the chef data and `pending_node_actions` is used to ensure consistency, and is worth doing if we have the chef data)

I would tend to go with the latter (the consistency is worth it, and I'm nearly sure we load the chef data in `apply_role_pre_chef_call` anyhow), but would be fine with the former.

Reply:
Btw, whatever option you pick, please add a comment to the code explaining the impact (cost of loading chef data vs consistency)

Reply:
umm, Im nearly sure that we added this in this place to avoid having a wrong `pending_node_actions` with a node that then its not included on the `batch`. Although I dont remember if we tested setting it below the pending_node_actions, but I dont think it should matter much, as it would only affect nodes that have not changed attributes or roles we dont need to re-save the role list to them (thus speeding up time as well?)

we tested this extensively adding roles, removing roles and moving a node between roles and it look like it worked as expected.

@ilausuch do you remember our thinking process for setting this in here?